# Problem definition

In order to evaluate PySpark framework, we want to perform an entire pipeline of steps, on a large dataset, and compare the same analysis using an usual framework, like scikit-learn, without paralel computation.

We will work on a classification problem, based on flights informations from North America. This dataset is available in Databricks datasets, and it contains 1,391,578 rows and 5 columns. It has no null values.

Our workflow will consist of feature engineering, cross-validation, hyperparameter tuning and simple versus ensemble model.

In [0]:
# Import libraries
import time
from pyspark.sql.functions import concat, hour, lit, to_timestamp
from pyspark.ml import Pipeline
from pyspark.ml.feature import Bucketizer, StringIndexer, OneHotEncoder, VectorAssembler
from pyspark.ml.classification import LogisticRegression, RandomForestClassifier
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator
from pyspark.ml.evaluation import BinaryClassificationEvaluator

# 1. Feature Engineering

Since PySpark Machine Learning algorithms require only numeric variables, we need to perform some transformations on the data.

First we will create a new column named label, representing our target variable, delayed flight or not.<br>
After that, we will implement steps of feature engineering to our pipeline.<br>
We will be transforming the date column into readable datetime format and bucketizing it into intervals of 3 hours along 24 hours.<br>
We will also be implementing an indexer for the categorical features (origin and destinantion) and one-hot encoder for them.

## 1.1. Target variable label

The Federal Aviation Administration (FAA) considers a flight to be "delayed" when it arrives 15 minutes or more after its scheduled time. Thus we will be creating the boolean label, accordingly to FAA definition. After that we will convert the boolean column into numerical.

In [0]:
data = spark.read.csv('dbfs:/databricks-datasets/flights/departuredelays.csv', header=True, inferSchema=True)

In [0]:
df = data
df = df.withColumn('label', df.delay >= 15)
df.groupBy('label').count().show()

+-----+-------+
|label|  count|
+-----+-------+
| true| 314474|
|false|1077104|
+-----+-------+



We can see that the data is unbalanced, having 3 times more non delayed flights.<br>
However we will work with the data without changing its distribution, in order to evaluate both models performance.

In [0]:
# Transform boolean into numerical
df = df.withColumn('label', df.label.cast('integer'))

In [0]:
df.display()

date delay distance origin destination label 1011245 6 602 ABE ATL 0 1020600 -8 369 ABE DTW 0 1021245 -2 602 ABE ATL 0 1020605 -4 602 ABE ATL 0 1031245 -4 602 ABE ATL 0 1030605 0 602 ABE ATL 0 1041243 10 602 ABE ATL 0 1040605 28 602 ABE ATL 1 1051245 88 602 ABE ATL 1 1050605 9 602 ABE ATL 0 1061215 -6 602 ABE ATL 0 1061725 69 602 ABE ATL 1 1061230 0 369 ABE DTW 0 1060625 -3 602 ABE ATL 0 1070600 0 369 ABE DTW 0 1071725 0 602 ABE ATL 0 1071230 0 369 ABE DTW 0 1070625 0 602 ABE ATL 0 1071219 0 569 ABE ORD 0 1080600 0 369 ABE DTW 0 1081230 33 369 ABE DTW 1 1080625 1 602 ABE ATL 0 1080607 5 569 ABE ORD 0 1081219 54 569 ABE ORD 1 1091215 43 602 ABE ATL 1 1090600 151 369 ABE DTW 1 1091725 0 602 ABE ATL 0 1091230 -4 369 ABE DTW 0 1090625 8 602 ABE ATL 0 1091219 83 569 ABE ORD 1 1101215 -5 602 ABE ATL 0 1100600 -5 369 ABE DTW 0 1101725 7 602 ABE ATL 0 1101230 -8 369 ABE DTW 0 1100625 52 602 ABE ATL 1 1101219 0 569 ABE ORD 0 1111215 127 602 ABE ATL 1 1110600 -9 369 ABE DTW 0 1110625 -4 602 ABE ATL 0 1121215 -5 602 ABE ATL 0 1121725 -1 602 ABE ATL 0 1131215 14 602 ABE ATL 0 1130600 -7 369 ABE DTW 0 1131725 -6 602 ABE ATL 0 1131230 -13 369 ABE DTW 0 1130625 29 602 ABE ATL 1 1131219 -8 569 ABE ORD 0 1140600 -9 369 ABE DTW 0 1141725 -9 602 ABE ATL 0 1141230 -8 369 ABE DTW 0 1140625 -5 602 ABE ATL 0 1141219 -10 569 ABE ORD 0 1150600 0 369 ABE DTW 0 1151725 -6 602 ABE ATL 0 1151230 0 369 ABE DTW 0 1150625 0 602 ABE ATL 0 1150607 0 569 ABE ORD 0 1151219 0 569 ABE ORD 0 1161215 -10 602 ABE ATL 0 1160600 -1 369 ABE DTW 0 1161725 -6 602 ABE ATL 0 1161230 -7 369 ABE DTW 0 1160625 -4 602 ABE ATL 0 1161219 68 569 ABE ORD 1 1171215 -8 602 ABE ATL 0 1170600 -5 369 ABE DTW 0 1171725 5 602 ABE ATL 0 1171230 -10 369 ABE DTW 0 1170625 -6 602 ABE ATL 0 1171219 -10 569 ABE ORD 0 1181215 -12 602 ABE ATL 0 1180600 -13 369 ABE DTW 0 1180625 0 602 ABE ATL 0 1191215 -16 602 ABE ATL 0 1191725 -5 602 ABE ATL 0 1201215 -8 602 ABE ATL 0 1201725 -5 602 ABE ATL 0 1201230 -11 369 ABE DTW 0 1200625 -7 602 ABE ATL 0 1201219 -6 569 ABE ORD 0 1210600 89 369 ABE DTW 1 1211725 0 602 ABE ATL 0 1211230 44 369 ABE DTW 1 1210625 -6 602 ABE ATL 0 1211219 9 569 ABE ORD 0 1220600 80 369 ABE DTW 1 1221230 -5 369 ABE DTW 0 1220625 333 602 ABE ATL 1 1220607 219 569 ABE ORD 1 1221219 15 569 ABE ORD 1 1231215 -12 602 ABE ATL 0 1230600 -3 369 ABE DTW 0 1231725 180 602 ABE ATL 1 1231230 -5 369 ABE DTW 0 1230625 -8 602 ABE ATL 0 1231219 -13 569 ABE ORD 0 1241215 -11 602 ABE ATL 0 1240600 -3 369 ABE DTW 0 1241725 2 602 ABE ATL 0 1241230 -5 369 ABE DTW 0 1240625 -3 602 ABE ATL 0 1241219 0 569 ABE ORD 0 1251215 2 602 ABE ATL 0 1250600 -4 369 ABE DTW 0 1250625 -4 602 ABE ATL 0 1261215 1 602 ABE ATL 0 1261725 7 602 ABE ATL 0 1271215 -4 602 ABE ATL 0 1270600 -9 369 ABE DTW 0 1271725 -8 602 ABE ATL 0 1271230 -7 369 ABE DTW 0 1270625 -1 602 ABE ATL 0 1271219 0 569 ABE ORD 0 1280600 -6 369 ABE DTW 0 1281725 -3 602 ABE ATL 0 1281230 -8 369 ABE DTW 0 1280625 -4 602 ABE ATL 0 1281219 0 569 ABE ORD 0 1290600 0 369 ABE DTW 0 1291230 0 369 ABE DTW 0 1290625 4 602 ABE ATL 0 1290607 92 569 ABE ORD 1 1291219 -10 569 ABE ORD 0 1301215 0 602 ABE ATL 0 1300600 -1 369 ABE DTW 0 1301725 10 602 ABE ATL 0 1301230 0 369 ABE DTW 0 1300625 0 602 ABE ATL 0 1301219 -11 569 ABE ORD 0 1311215 1 602 ABE ATL 0 1310600 -4 369 ABE DTW 0 1311725 25 602 ABE ATL 1 1311230 -4 369 ABE DTW 0 1310625 -5 602 ABE ATL 0 1311219 24 569 ABE ORD 1 1011400 26 137 ABI DFW 1 1011525 -11 137 ABI DFW 0 1021400 -5 137 ABI DFW 0 1021525 47 137 ABI DFW 1 1031400 -12 137 ABI DFW 0 1031525 40 137 ABI DFW 1 1041400 -12 137 ABI DFW 0 1041525 3 137 ABI DFW 0 1051400 -5 137 ABI DFW 0 1051525 -5 137 ABI DFW 0 1061400 2 137 ABI DFW 0 1061525 22 137 ABI DFW 1 1071400 -12 137 ABI DFW 0 1071525 -15 137 ABI DFW 0 1081525 21 137 ABI DFW 1 1091525 -8 137 ABI DFW 0 1101525 4 137 ABI DFW 0 1111525 -23 137 ABI DFW 0 1121525 -9 137 ABI DFW 0 1131525 -12 137 ABI DFW 0 1141525 -5 137 ABI DFW 0 1151525 -17 137 ABI DFW 0 1161525 2 137 ABI DFW 0 1171525 -9 137 ABI DF

In [0]:
# Get number of records
print("The data contain %d records." % df.count())

The data contain 1391578 records.


In [0]:
# Print schema of DataFrame
df.printSchema()

root
 |-- date: integer (nullable = true)
 |-- delay: integer (nullable = true)
 |-- distance: integer (nullable = true)
 |-- origin: string (nullable = true)
 |-- destination: string (nullable = true)
 |-- label: integer (nullable = true)



## 1.2. Column departure

We will first convert the time column into string, so next we can convert it into timestamp.<br>
Then we will declare the bucketizer step for intervals of 3 hours.

In [0]:
# Transform int to string
df = df.withColumn('departure', concat(lit('20140'), df.date.cast('string')))

# Transform string to timestamp
df = df.withColumn('departure', to_timestamp('departure', 'yyyyMMddHHmm'))

# Get hour from departure time
df = df.withColumn('hour', hour(df.departure))

In [0]:
# Define Bucketizer
bucketizer = Bucketizer(splits=[0,3,6,9,12,15,18,21,24], inputCol='hour', outputCol='departure_bucket')

## 1.3. Categorical features

The categorical features origin and destination contain the IATA code for airports of North America. Since there are around 300 different airports in the dataset, we will replace them by their state, reducing to a total of 65 states.

In order to do this, we will need to import a second database, containing the airports informations. We will perform a join between the two tables, to get the corresponding states.

After this, we will implement an indexer and a one-hot encoder for each categorical feature.<br>
The buckets from previous section will need one-hot encoding also, since they are categorical values now.<br>
These steps will be gathered in the pipeline.

In [0]:
df_air = spark.read.option("header", "true").option("delimiter", "\t").csv('dbfs:/databricks-datasets/flights/airport-codes-na.txt')

In [0]:
df_air.display()

City State Country IATA Abbotsford BC Canada YXX Aberdeen SD USA ABR Abilene TX USA ABI Akron OH USA CAK Alamosa CO USA ALS Albany GA USA ABY Albany NY USA ALB Albuquerque NM USA ABQ Alexandria LA USA AEX Allentown PA USA ABE Alliance NE USA AIA Alpena MI USA APN Altoona PA USA AOO Amarillo TX USA AMA Anahim Lake BC Canada YAA Anchorage AK USA ANC Appleton WI USA ATW Arviat NWT Canada YEK Asheville NC USA AVL Aspen CO USA ASE Athens GA USA AHN Atlanta GA USA ATL Atlantic City NJ USA ACY Augusta GA USA AGS Augusta ME USA AUG Austin TX USA AUS Bagotville PQ Canada YBG Baie-Comeau PQ Canada YBC Bakersfield CA USA BFL Baltimore MD USA BWI Bangor ME USA BGR Bar Harbor ME USA BHB Barrow AK USA BRW Baton Rouge LA USA BTR Beaumont TX USA BPT Beckley WV USA BKW Bedford MA USA BED Bellingham WA USA BLI Bemidji MN USA BJI Bethel AK USA BET Bettles AK USA BTT Billings MT USA BIL Binghamton NY USA BGM Birmingham AL USA BHM Bismarck ND USA BIS Bloomington IL USA BMI Bloomington IN USA BMG Bluefield WV USA BLF Boise ID USA BOI Boston MA USA BOS Bozeman MT USA BZN Brookings SD USA BKX Brownsville TX USA BRO Brunswick GA USA BQK Buffalo NY USA BUF Burbank CA USA BUR Burlington IA USA BRL Burlington MA USA BBF Burlington VT USA BTV Butte MT USA BTM Calgary AB Canada YYC Campbell River BC Canada YBL Cap-aux-Meules, Magdalens Islands PQ Canada YGR Cape Girardeau MO USA CGI Carlsbad CA USA CLD Carlsbad NM USA CNM Casper WY USA CPR Castlegar BC Canada YCG Cedar Rapids IA USA CID Champaign IL USA CMI Charleston SC USA CHS Charleston WV USA CRW Charlotte NC USA CLT Charlottesville VA USA CHO Charlottetown PEI Canada YYG Chattanooga TN USA CHA Cheyenne WY USA CYS Chibougamau PQ Canada YMT Chicago IL USA MDW Chicago IL USA ORD Chicago IL USA CHI Chico CA USA CIC Churchill MB Canada YYQ Cincinnati OH USA CVG Clarksburg WV USA CKB Cleveland OH USA CLE Clovis NM USA CVN Cody WY USA COD College Station TX USA CLL Colorado Springs CO USA COS Columbia MO USA COU Columbia SC USA CAE Columbus GA USA CSG Columbus IN USA CLU Columbus MS USA GTR Columbus NE USA OLU Columbus OH USA CMH Cordova AK USA CDV Corpus Christi TX USA CRP Cranbrook BC Canada YXC Dallas TX USA DFW Dallas TX USA DAL Dayton OH USA DAY Daytona Beach FL USA DAB Decatur IL USA DEC Deer Lake NF Canada YDF Denver CO USA DEN Des Moines IA USA DSM Detroit MI USA DTW Detroit MI USA DTT Devils Lake ND USA DVL Dickinson ND USA DIK Dillingham AK USA DLG Dodge City KS USA DDC Dothan AL USA DHN Dryden ON Canada YHD Du Bois PA USA DUJ Dubuque IA USA DBQ Duluth MN USA DLH Durango CO USA DRO Dutch Harbor, Un Island AK USA DUT Eau Claire WI USA EAU Edmonton AB Canada YEG Eek AK USA EEK El Centro CA USA IPL El Dorado AR USA ELD El Paso TX USA ELP Elko NV USA EKO Elmira NY USA ELM Enid OK USA WDG Erie PA USA ERI Escanaba MI USA ESC Eugene OR USA EUG Eureka CA USA ACV Evansville IN USA EVV Fairbanks AK USA FAI Fargo ND USA FAR Farmington NM USA FMN Fayetteville AR USA XNA Fayetteville NC USA FAY Flagstaff AZ USA FLG Flin Flon MB Canada YFO Flint MI USA FNT Florence SC USA FLO Fort Dodge IA USA FOD Fort Lauderdale FL USA FLL Fort Leonard Wood MO USA TBN Fort McMurray AB Canada YMM Fort Myers FL USA RSW Fort Nelson BC Canada YYE Fort Saint John BC Canada YXJ Fort Smith AR USA FSM Fort Smith NWT Canada YSM Fort Walton Beach FL USA VPS Fort Wayne IN USA FWA Fort Yukon AK USA FYU Fredericton NB Canada YFC Fresno CA USA FAT Gainesville FL USA GNV Gander NF Canada YQX Garden City KS USA GCK Gaspe PQ Canada YGP Gillette WY USA GCC Glendive MT USA GDV Grand Forks ND USA GFK Grand Island NE USA GRI Grand Junction CO USA GJT Grand Rapids MI USA GRR Grande Prairie AB Canada YQU Great Bend KS USA GBD Great Falls MT USA GTF Green Bay WI USA GRB Greenbrier WV USA LWB Greensboro NC USA GSO Greenville MS USA GLH Greenville NC USA PGV Greenville SC USA GSP Gulfport MS USA GPT Gunnison CO USA GUC Hagerstown MD USA HGR Halifax NS Canada YHZ Hamilton ON Canada YHM Hana, Maui HI USA HNM Hancock MI USA CMX Harlingen TX USA HRL Harrisb

In [0]:
# Join on origin
df_origin = df_air.withColumnRenamed('IATA', 'origin')
df_origin = df_origin.withColumnRenamed('State', 'origin_state')
df_origin = df_origin.select('origin_state', 'origin')
df = df.join(df_origin, on='origin', how='leftouter')

In [0]:
# Join on destination
df_dest = df_air.withColumnRenamed('IATA', 'destination')
df_dest = df_dest.withColumnRenamed('State', 'dest_state')
df_dest = df_dest.select('dest_state', 'destination')
df = df.join(df_dest, on='destination', how='leftouter')

In [0]:
df = df.dropna()

In [0]:
# Define Indexer
indexer = StringIndexer(inputCols=['origin_state', 'dest_state'],
                        outputCols=['origin_index', 'dest_index'])

In [0]:
# Define One-Hot Encoder
encoder = OneHotEncoder(inputCols=['departure_bucket', 'origin_index', 'dest_index'],
                        outputCols=['departure_dummy', 'origin_dummy', 'dest_dummy'])

## 1.4. Train-test split

Before applying any fitting or prediction, we will split the data into training and test sets. This will ensure that data leakage does not occur during all process.

In [0]:
# Split the data
train, test = df.randomSplit([.7, .3])

# 2. Pipeline

Now we will define a pipeline to run all previous declared steps.

## 2.1. Vector Assembler

This step is required for PySpark, since it requires all features into one vector, in sparse representation.

In [0]:
# Assembling columns
assembler = VectorAssembler(inputCols=['delay', 'distance', 'departure_dummy', 'origin_dummy', 'dest_dummy'],
                            outputCol='features')

## 2.2. Pipeline

Now we can declare and build our pipeline with all previous steps.<br>
By applying the pipeline, we can properly fit with the training data and transform both the training and test sets.

In [0]:
# Create the Pipeline
pipeline = Pipeline(stages=[bucketizer, indexer, encoder, assembler])

In [0]:
pipeline = pipeline.fit(train)

In [0]:
train = pipeline.transform(train)

In [0]:
test = pipeline.transform(test)

In [0]:
# Displaying the training dataset after the pipeline
train.display()

destination origin date delay distance label departure hour origin_state dest_state departure_bucket origin_index dest_index departure_dummy origin_dummy dest_dummy features ABE ATL 1012039 -4 602 0 2014-01-01T20:39:00.000+0000 20 GA PA 6.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(6), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(0, 1, 8, 12, 69), values -> List(-4.0, 602.0, 1.0, 1.0, 1.0)) ABE ATL 1021016 -7 602 0 2014-01-02T10:16:00.000+0000 10 GA PA 3.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(0, 1, 5, 12, 69), values -> List(-7.0, 602.0, 1.0, 1.0, 1.0)) ABE ATL 1022039 0 602 0 2014-01-02T20:39:00.000+0000 20 GA PA 6.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(6), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(1, 8, 12, 69), values -> List(602.0, 1.0, 1.0, 1.0)) ABE ATL 1032039 0 602 0 2014-01-03T20:39:00.000+0000 20 GA PA 6.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(6), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(1, 8, 12, 69), values -> List(602.0, 1.0, 1.0, 1.0)) ABE ATL 1041015 -3 602 0 2014-01-04T10:15:00.000+0000 10 GA PA 3.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(0, 1, 5, 12, 69), values -> List(-3.0, 602.0, 1.0, 1.0, 1.0)) ABE ATL 1051016 84 602 1 2014-01-05T10:16:00.000+0000 10 GA PA 3.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(0, 1, 5, 12, 69), values -> List(84.0, 602.0, 1.0, 1.0, 1.0)) ABE ATL 1060930 -4 602 0 2014-01-06T09:30:00.000+0000 9 GA PA 3.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(0, 1, 5, 12, 69), values -> List(-4.0, 602.0, 1.0, 1.0, 1.0)) ABE ATL 1061502 82 602 1 2014-01-06T15:02:00.000+0000 15 GA PA 5.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(5), values -> List(1.0)) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, length -> 106, indices -> List(0, 1, 7, 12, 69), values -> List(82.0, 602.0, 1.0, 1.0, 1.0)) ABE ATL 1062115 0 602 0 2014-01-06T21:15:00.000+0000 21 GA PA 7.0 3.0 12.0 Map(vectorType -> sparse, length -> 7, indices -> List(), values -> List()) Map(vectorType -> sparse, length -> 48, indices -> List(3), values -> List(1.0)) Map(vectorType -> sparse, length -> 49, indices -> List(12), values -> List(1.0)) Map(vectorType -> sparse, len

# 3. Baseline Model

We will apply two classification models: one simple logistic regression and one ensemble method, Random Forest.<br>
Since the logistic regression is more simple, we will use it as baseline model.<br>
We will perform hyperparameter tuning on it, with grid-search cross-validation and evaluate the best model on the test set.<br>
We will apply two evaluations for both models: ROC-AUC and Confusion Matrix.

## 3.1. Evaluators

Since we will be evaluating both models with the same roc_auc metric, we will define it first.<br>
We will also define a function to calculate the Confusion Matrix as well as accuracy, precision, recall and F1-Score.

In [0]:
# Define the evaluator
evaluator = BinaryClassificationEvaluator(metricName='areaUnderROC')

In [0]:
# Confusion Matrix
def confusion_matrix(prediction):
  prediction.groupBy('label', 'prediction').count().show()
  
  # Calculate the elements of the confusion matrix
  TN = prediction.filter('prediction = 0 AND label = prediction').count()
  TP = prediction.filter('prediction = 1 AND label = prediction').count()
  FN = prediction.filter('prediction = 0 AND label <> prediction').count()
  FP = prediction.filter('prediction = 1 AND label <> prediction').count()
  
  # Accuracy, precision and recall metrics
  accuracy = (TP+TN)/(TP+TN+FP+FN)
  precision = TP/(TP+FP)
  recall = TP/(TP+FN)
  f1 = 2*(precision*recall)/(precision+recall)
  
  print('Accuracy: ', accuracy)
  print('Precision: ', precision)
  print('Recall: ', recall)
  print('F1-Score: ', f1)

## 3.2. Logistic Regression

The Logistic Regression will be our baseline model, for comparison.<br>
For its hyperparameters we will search for 'regParam' wich is the Regularization λ and 'elasticNetParam' which is the Regularization α (L1/L2).

In [0]:
# Create Logistic Regression model
lr = LogisticRegression()

In [0]:
# Make a grid for grid-search
params_lr = ParamGridBuilder()

# Add the hyperparameters
params_lr = params_lr.addGrid(lr.regParam, [1, .1, .01])
params_lr = params_lr.addGrid(lr.elasticNetParam, [0, 1])

# Build the grid
params_lr = params_lr.build()

## 3.3. Hyperparameter Tuning

Now through grid-search cross-validation, we may search for the best hyperparameters, on the training set, and use the best model to predict the test set, in order to evaluate the results.

In [0]:
# Create the CrossValidator
cv_lr = CrossValidator(estimator=lr,
                       estimatorParamMaps=params_lr,
                       evaluator=evaluator,
                       numFolds=3)

In [0]:
# Train the model and time it
t_start = time.time()
cv_lr = cv_lr.fit(train)
t_total = time.time() - t_start
print('Cross-validation training with Logistic Regression took around {:.2f} min.'.format(t_total/60))

Cross-validation training with Logistic Regression took around 7.91 min.


## 3.4. Cross-Validation Best Model

In [0]:
# Extract the best model
best_lr = cv_lr.bestModel

# Print best_lr
print(best_lr)

LogisticRegressionModel: uid=LogisticRegression_34e3dd8a08f2, numClasses=2, numFeatures=106


In [0]:
best_lr.explainParam('regParam')

Out[77]: 'regParam: regularization parameter (>= 0). (default: 0.0, current: 0.1)'

In [0]:
best_lr.explainParam('elasticNetParam')

Out[78]: 'elasticNetParam: the ElasticNet mixing parameter, in range [0, 1]. For alpha = 0, the penalty is an L2 penalty. For alpha = 1, it is an L1 penalty. (default: 0.0, current: 1.0)'

## 3.5. Model Evaluation

In [0]:
# Testing the data
predictions_lr = cv_lr.transform(test)

In [0]:
# Evaluate the predictions
print('roc_auc score: ', evaluator.evaluate(predictions_lr))

confusion_matrix(predictions_lr)

roc_auc score:  1.0
+-----+----------+------+
|label|prediction| count|
+-----+----------+------+
|    1|       0.0| 54001|
|    0|       0.0|300984|
|    1|       1.0| 34712|
+-----+----------+------+

Accuracy:  0.8614282378360624
Precision:  1.0
Recall:  0.3912842537170426
F1-Score:  0.562479238403889


# 4. Ensemble Random Forest

Now we will apply a Random Forest model, also searching for its best hyperparameters, through grid-search cross-validation.

## 4.1. Model

For its hyperparameters we will search for 'featureSubsetStrategy' wich is the number of features to consider and 'maxDepth' which is the maximum depth of the tree.

In [0]:
# Create Random Forest model
rf = RandomForestClassifier()

In [0]:
# Make a grid for grid-search
params_rf = ParamGridBuilder()

# Add the hyperparameters
params_rf = params_rf.addGrid(rf.featureSubsetStrategy, ['all', 'onethird', 'sqrt', 'log2'])
params_rf = params_rf.addGrid(rf.maxDepth, [2, 5, 10])

# Build the grid
params_rf = params_rf.build()

## 4.2. Hyperparameter Tuning

In [0]:
# Create the CrossValidator
cv_rf = CrossValidator(estimator=rf,
                       estimatorParamMaps=params_rf,
                       evaluator=evaluator,
                       numFolds=3)

In [0]:
# Train the model and time it
t_start = time.time()
cv_rf = cv_rf.fit(train)
t_total = time.time() - t_start
print('Cross-validation training with Random Forest took around {:.2f} min.'.format(t_total/60))

Cross-validation training with Random Forest took around 22.73 min.


## 4.3. Cross-Validation Best Model

In [0]:
# Extract the best model
best_rf = cv_rf.bestModel

# Print best_rf
print(best_rf)

RandomForestClassificationModel: uid=RandomForestClassifier_2d8266ad910a, numTrees=20, numClasses=2, numFeatures=106


In [0]:
best_rf.explainParam('featureSubsetStrategy')

Out[86]: "featureSubsetStrategy: The number of features to consider for splits at each tree node. Supported options: 'auto' (choose automatically for task: If numTrees == 1, set to 'all'. If numTrees > 1 (forest), set to 'sqrt' for classification and to 'onethird' for regression), 'all' (use all features), 'onethird' (use 1/3 of the features), 'sqrt' (use sqrt(number of features)), 'log2' (use log2(number of features)), 'n' (when n is in the range (0, 1.0], use n * number of features. When n is in the range (1, number of features), use n features). default = 'auto' (default: auto, current: all)"

In [0]:
best_rf.explainParam('maxDepth')

Out[87]: 'maxDepth: Maximum depth of the tree. (>= 0) E.g., depth 0 means 1 leaf node; depth 1 means 1 internal node + 2 leaf nodes. Must be in range [0, 30]. (default: 5, current: 5)'

## 4.4. Model Evaluation

In [0]:
# Testing the data
predictions_rf = cv_rf.transform(test)

In [0]:
# Evaluate the predictions
print('roc_auc score: ', evaluator.evaluate(predictions_rf))

confusion_matrix(predictions_rf)

roc_auc score:  0.999801222799958
+-----+----------+------+
|label|prediction| count|
+-----+----------+------+
|    1|       0.0|  2685|
|    0|       0.0|300416|
|    1|       1.0| 86028|
|    0|       1.0|   568|
+-----+----------+------+

Accuracy:  0.991652488985032
Precision:  0.9934408055799344
Recall:  0.9697338608772108
F1-Score:  0.98144419282524


# 5. Conclusions

Spark is a framework developed specially for big data processing.<br>
Its default HDFS partition size is 64MB, which means for larger data, Spark process with multiple processors. This way, the processing time does not scalate with the data size.

The pyspark.ml is the Machine Learning library, with Python interface, developed to work properly with all the advantages of the Spark framework.

Analyzing both models evaluation, we can see that, although Random Forest achieved a slightly lower ROC AUC Score, its Confusion Matrix metrics were much better than Logistic Regression.

|Metric|Logistic Regression|Random Forest|
|--|--|--|
|ROC AUC|1.0000|0.9998|
|Accuracy|0.8614|0.9917|
|Precision|1.0000|0.9934|
|Recall|0.3913|0.9697|
|F1-Score|0.5625|0.9814|

The most significant difference is the high Recall and F1-Score for Random Forest, meaning that the Ensemble Model is capable of achieving better predictions of true class (delay). Other diffence is that Logistic Regression is a linear model, while Decision Tree Classifiers are capable of modeling non-linearities from the data.

The unbalanced distribution of the data affects both classification models. However since Random Forest is an ensemble model, composed of multiple different predictors, it was capable of compensate the unbalance. This unbalance was also responsible for Logistic Regression's high accuracy and precision.

For the entire data of 1.4 million rows, the Logistic Regression cross-validation took around 8 min to run, while the Random Forest took around 22 min.